### ISSUE TITLE
"raw" image is wrongly formed

### ISSUE THREAD

**Issue by olivier-haddad** [christianrauch/raspicam2_node#11]
I configured the raspicam2 node with "raw" image transport and 1640x1232 resolution, something that I used with ros1 with the same camera.

The output of raspicam2 node window:
```
[INFO] [raspicam2_node-1]: process started with pid [10078]
[raspicam2_node-1] mmal: Argument is invalid
[raspicam2_node-1] Camera component done
[raspicam2_node-1] Image encoder component done
[raspicam2_node-1] Video encoder component done
[raspicam2_node-1] splitter component done
[raspicam2_node-1] Starting video capture (1640, 1232, 80, 30)
[raspicam2_node-1] Video capture started

```

I tried both with my recipient node and rqt and both show the following message:

`cv_bridge exception: Image is wrongly formed: height * step != size  or  0 * 0 != 6091008
`

Usually, this message appears with height and step values. Here, they're shown both as 0.

If raspicam2 is configured with the "compressed" image transport, a black image is displayed in rqt.

raspivid works so this is not a wrongly installed device or bad drivers.

**Comment by christianrauch** [comment]
I can confirm this on different resolutions and framerates. I don't have a clue what causes this. I haven't used the camera module for a long time now. Do you know if the ROS1 node is working in your configuration? I updated the driver a while ago to pull changes from the original ROS1 project (94e968e4598568e98839211ea7418844aa8b707e). This might have broken the driver.

**Comment by olivier-haddad** [comment]
In ROS1, I used quite an old version (0.4.0 nov 2018 [06428f9](https://github.com/UbiquityRobotics/raspicam_node/commit/06428f97b0fc6f90c1093cdc54f9b0032099a9df)) of the ubiquity repo and it was working fine.
When did you last test your ROS2 package? In which commit?

**Comment by olivier-haddad** [comment]
This bug still exists in the ubiquity version and may be related to an invalid stride (step).
In the ROS1 version, I fixed this bug by calculating the step that must be a multiple of 16 for video capture ([see here](https://picamera.readthedocs.io/en/release-1.13/recipes2.html#array-capture)).
So that I turned this:
```
    image.msg.step = (pData->pstate.width * 3);
```
into that:
```
    uint32_t rounded_width = (uint32_t)(ceil(pData->pstate.width/16.0) * 16);//aligned to 16
    image.msg.step = rounded_width * 3 ; 
```
I am trying to fix the same thing in your code but you refactored quite extensively so that I am not sure that you are not setting the values elsewhere (I'll have to check).
In RasPiCamPublisherNode.cpp, in onImageRaw, I added the following:
```
    msg->height = camera_info.height;
    msg->width = camera_info.width;
    //calculating the step, must be for video capture a multiple of 16 (32 for stills)
    uint32_t multiples16 = msg->width / 16u;
    if(msg->width % 16 != 0)
    	multiples16++;
    msg->step = multiples16 * 16 * 3;
    msg->encoding = "bgr8";

```
I also noticed that you set the encoding to rgb8 instead of bgr8 in the original ubiquity code.
Did you select rgb8 for the actual capture or perform  a conversion somewhere else or is there another reason?

**Comment by olivier-haddad** [comment]
After the corrections above, there are no error messages anymore but the images are blank.
Debugging in gdb and some old fashioned traces showed that all the pixels have a zero value when returned in the raspicam.cpp:splitter_buffer_callback method.

**Comment by svengit** [comment]
Any news on this? This is a hard blocker for a project I am working on.

**Comment by christianrauch** [comment]
I am not actively using the node at the moment. Unfortunately, there is no common "raspicam" library that the ROS1 and ROS2 nodes could share. Porting the features introduced in the ROS1 node to ROS2 is therefore very time consuming and error prone.

The quickest "solution" for now would be to checkout an older version from the git history. My guess is that this issue was introduced with merging the upstream changes in f2e83b40697bb3488bdaef8172c4e6300fb1fff4.
Checkout the commit before this (d29734b2e4083a717e36a4154913d94a5f860d5d) and see if you can compile the node with the old ROS2 API.

**Comment by svengit** [comment]
Ok, thanks for your quick reply. Alternatively I'll try to use the ROS1 node in combination with the bridge.

**Comment by christianrauch** [comment]
Alternatively, you can use the V4L2 driver for the raspberry pi camera (`https://www.raspberrypi.org/forums/viewtopic.php?t=62364`, `https://github.com/cmcmurrough/teaching/blob/master/raspi_notes/v4l2 raspicam install.txt`) and then use the generic V4L2 ROS2 node (https://gitlab.com/boldhearts/ros2_v4l2_camera).

**Comment by christianrauch** [comment]
> After the corrections above, there are no error messages anymore but the images are blank.

This should be fixed by 82603fd8935dd91586feb1e42011302043db41d6. The image did appear as "empty" because the default brightness was set to 0. In fact, the image just used a very low brightness that resulted in very low intensity values (<10). You can see this when you visualise the image with a different colour mapping, e.g. in GIMP with thresholds.
